core/market/decentralized/src/negotiation/requestor.rs

```diff
@@ -43,7 +54,132 @@ impl RequestorNegotiationEngine {
         &self,
         subscription_id: &SubscriptionId,
     ) -> Result<(), NegotiationError> {
-        // TODO: Implement
+        self.notifier.stop_notifying(subscription_id).await;
+
+        // We can ignore error, if removing events failed, because they will be never
+        // queried again and don't collide with other subscriptions.
+        let _ = self
+            .db
+            .as_dao::<EventsDao>()
+            .remove_requestor_events(subscription_id)
+            .await
+            .map_err(|e| {
+                log::warn!(
+                    "Failed to remove events related to subscription [{}].",
+                    subscription_id
+                )
+            });
+        // TODO: We could remove all resources related to Proposals.
         Ok(())
     }
+
+    pub async fn query_events(
+        &self,
+        subscription_id: &SubscriptionId,
+        timeout: f32,
+        max_events: Option<i32>,
+    ) -> Result<Vec<RequestorEvent>, QueryEventsError> {
+        let mut timeout = Duration::from_millis((1000.0 * timeout) as u64);
+        let stop_time = Instant::now() + timeout;
+        let max_events = max_events.unwrap_or(i32::max_value());
+
+        if max_events < 0 {
+            Err(QueryEventsError::InvalidMaxEvents(max_events))?
+        } else if max_events == 0 {
+            return Ok(vec![]);
+        }
+
+        loop {
+            let events = get_events_from_db(&self.db, subscription_id, max_events).await?;
+            if events.len() > 0 {
+                return Ok(events);
+            }
+
+            // Solves panic 'supplied instant is later than self'.
+            if stop_time < Instant::now() {
```

Comment:
it might be clearer stated when you use `while` with this condition instead of `loop`

Reply:
I would prefer to do it this way. But it is required to check events in database at least once. If this would be loop condition, we could not enter loop at all, so I would have to duplicate checking db before loop. And it doesn't look so good as we would expect